File: python/algo/202504/algo1547.py
from cmath import inf
from functools import cache
from typing import List

# 动态规划 -  8.2其他区间 DP
class Solution:
    def minCost(self, n: int, cuts: List[int]) -> int:
        # 自行解答，记忆化搜索 O(N^3)
        # cuts.sort()

        @cache
        def dfs(i:int, j:int) -> int:

            res = inf
            for c in cuts:
                if (i < c < j):
                    res = min(res, dfs(i, c) + dfs(c, j))
            if res == inf: return 0
            res += j - i
            return res
        return dfs(0, n)

File: cplus/leetcode/202504/algo1547.cpp
#include <bits/stdc++.h>
using namespace std;

// 动态规划 -  8.2其他区间 DP
class Solution {
public:
    int minCost(int n, vector<int>& cuts) {
        // 自行解答，记忆化搜索 O(N^3)
        
        unordered_map<long, int> memo;
        auto dfs = [&](this auto&& dfs, int i, int j) {
            long mask = (long) i << 32 | j;
            if (memo.count(mask)) return memo[mask];

            auto& res = memo[mask];
            res = INT_MAX;
            for (auto v : cuts) {
                if (i < v && v < j) {
                    res = min(res, dfs(i, v) + dfs(v, j));
                }
            }

            if (res == INT_MAX) return res = 0;
            return res += j - i;
        };

        return dfs(0, n);

    }
};